json_object.c

```diff
@@ -1125,71 +1232,65 @@ int json_object_set_double(struct json_object *jso, double new_value)
 static int json_object_string_to_json_string(struct json_object *jso, struct printbuf *pb,
                                              int level, int flags)
 {
+	ssize_t len = JC_STRING(jso)->len;
 	printbuf_strappend(pb, "\"");
-	json_escape_str(pb, get_string_component(jso), jso->o.c_string.len, flags);
+	json_escape_str(pb, get_string_component(jso), len < 0 ? -(ssize_t)len : len, flags);
 	printbuf_strappend(pb, "\"");
 	return 0;
 }
 
 static void json_object_string_delete(struct json_object *jso)
 {
-	if (jso->o.c_string.len >= LEN_DIRECT_STRING_DATA)
-		free(jso->o.c_string.str.ptr);
+	if (JC_STRING(jso)->len < 0)
+		free(JC_STRING(jso)->c_string.pdata);
 	json_object_generic_delete(jso);
 }
 
-struct json_object *json_object_new_string(const char *s)
+static struct json_object *_json_object_new_string(const char *s, const size_t len)
 {
-	struct json_object *jso = json_object_new(json_type_string);
+	size_t objsize;
+	struct json_object_string *jso;
+
+	/*
+     * Structures           Actual memory layout
+     * -------------------  --------------------
+	 * [json_object_string  [json_object_string
+	 *  [json_object]        [json_object]
+     *  ...other fields...   ...other fields...
+	 *  c_string]            len
+     *                       bytes
+	 *                       of
+	 *                       string
+	 *                       data
+     *                       \0]
+	 */
+	if (len > (SSIZE_T_MAX - (sizeof(*jso) - sizeof(jso->c_string)) - 1))
+		return NULL;
+	objsize = (sizeof(*jso) - sizeof(jso->c_string)) + len + 1;
+	if (len < sizeof(void *))
+		// We need a minimum size to support json_object_set_string() mutability
+		// so we can stuff a pointer into pdata :(
+		objsize += sizeof(void *) - len;
+
+    jso = (struct json_object_string *)json_object_new(json_type_string, objsize,
+		&json_object_string_to_json_string, &json_object_string_delete);
```

Comment:
But I am curious about how much parformance will be improved if the `LEN_DIRECT_STRING_DATA` 32 is set to 8-12 according to the previous logic.

Reply:
LEN_DIRECT_STRING_DATA is gone with these changes, so there in nothing to tune now.  In effect, *all* strings are direct now, with the amount allocated being the bare minimum needed.
If you want to check the performance of the code from master with LEN_DIRECT_STRING_DATA=8, vs. json_object-split code, you should be able to do that by making that change locally, then running something like `jc-bench.sh --before . --after json_object-split`

Reply:
d'oh, jc-bench.sh isn't quite working.  hmm... I'll need to fix something with it.

Reply:
Actually jc-bench.sh is fine, just need to tweak how to run it a bit: `./jc-bench.sh --before origin/json_object-split`
With LEN_DIRECT_STRING_DATA=8 on the master branch, the decrease in memory usage is not quite as much (only 9% decrease vs. 12% decrease), which is expected.
The runtime improvement drops from 5.8% better to only 0.5% better.  I'm not sure why using 8 vs 32 makes the performance of master that much better all by itself, but even so the json_object-split branch still comes out (barely) ahead.

Reply:
ok, I get it